audiostream/src/ni/media/audio/source/gstreamer_file_source.cpp

```diff
@@ -0,0 +1,337 @@
+
+void gstreamer_file_source::init_gstreamer()
+{
+    static bool gstreamer_initialized = false;
+
+    if ( !std::exchange( gstreamer_initialized, true ) )
+    {
+        gst_init( 0, nullptr );
+    }
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+void gstreamer_file_source::setup_source( const std::string& path, audio::ifstream_info::container_type container )
+{
+    GstElement* sink        = prepare_pipeline( path );
+    auto        sinkpad     = gst_element_get_static_pad( sink, "sink" );
+    auto        caps        = gst_pad_get_current_caps( sinkpad );
+    auto        caps_struct = gst_caps_get_structure( caps, 0 );
+    fill_format_info( caps_struct, container );
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+GstElement* gstreamer_file_source::prepare_pipeline( const std::string& path )
+{
+    m_pipeline.reset( gst_pipeline_new( "pipeline" ) );
+
+    GstElement* source    = gst_element_factory_make( "filesrc", "source" );
+    GstElement* decodebin = gst_element_factory_make( "decodebin", "decoder" );
+    GstElement* queue     = gst_element_factory_make( "queue", "queue" );
+    GstElement* sink      = gst_element_factory_make( "appsink", "sink" );
+
+    gst_bin_add_many( GST_BIN( m_pipeline.get() ), source, decodebin, queue, sink, nullptr );
+    gst_element_link_many( source, decodebin, NULL );
+    gst_element_link_many( queue, sink, NULL );
+
+    g_object_set( source, "location", path.c_str(), NULL );
+
+    g_signal_connect( decodebin, "pad-added", G_CALLBACK( &onPadAdded ), queue );
+
+    preroll_pipeline();
+    return sink;
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+void gstreamer_file_source::preroll_pipeline()
+{
+    gst_element_set_state( m_pipeline.get(), GST_STATE_PAUSED );
+
+    if ( wait_for_async_operation() != GST_STATE_PAUSED )
+        throw std::runtime_error( "gstreamer_file_source: pipeline doesn't preroll into paused state" );
+
+    gst_element_set_state( m_pipeline.get(), GST_STATE_PLAYING );
+
+    if ( wait_for_async_operation() != GST_STATE_PLAYING )
+        throw std::runtime_error( "gstreamer_file_source: pipeline doesn't preroll into playing state" );
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+void gstreamer_file_source::fill_format_info( GstStructure*                        caps_struct,
+                                              audio::ifstream_info::container_type container )
+{
+    m_info.container( container );
+    m_info.codec( audio::ifstream_info::codec_type::mp3 );
+    m_info.lossless( false );
+
+    gint64 num_frames = 0;
+    if ( !gst_element_query_duration( m_pipeline.get(), GST_FORMAT_DEFAULT, &num_frames ) )
```

Comment:
Queries in `GST_FORMAT_DEFAULT` or often not possible. It would be safer to do a query in `TIME` and then convert

Reply:
Done.